File: 206-Reverse-Linked-List.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */
class Solution {
    /// Reverse a linked list
    /// Time Complexity: O(n)
    /// Space Complexity: O(n)
    ///
    /// - Parameters:
    ///   - head: a head node of a given linked list
    /// - Returns: a head node of a reversed given linked list
    func reverseList(_ head: ListNode?) -> ListNode? {
        // Create a node array of a given linked list
        var nodeStack: [ListNode] = []
        var currentPos: ListNode? = head
        while currentPos != nil {
            nodeStack.append(currentPos!)
            currentPos = currentPos?.next
        }

        // Keep first node in a result
        currentPos = nodeStack.popLast()
        var prevPos: ListNode? = currentPos
        let tailPos: ListNode? = currentPos

        // Switch next node and previous node
        while currentPos != nil {
            currentPos = nodeStack.popLast()
            prevPos?.next = currentPos
            prevPos = currentPos
        }

        // Set nil to last node next property
        prevPos?.next = nil

        // Return first node(last node of a given list)
        return tailPos
    }
}

// Sample Testcase:
//   Input:
//     head object of a linked list(1:1 -> 2:2 -> 3:3 -> 4:4 -> 5:5 -> nil)
//   Output:
//     a reversed list head(1:5 -> 2:4 -> 3:3 -> 4:2 -> 5:1 -> nil)

// Performance Result:
//   Runtime: 20 ms, faster than 63.54% of Swift online submissions for Reverse Linked List.
//   Memory Usage: 19.9 MB, less than 31.11% of Swift online submissions for Reverse Linked List.

File: 206-Reverse-Linked-List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if head is None:
            return None

        stuck = []
        curNode = head
        while curNode is not None:
            stuck.insert(0, curNode)
            curNode = curNode.next

        newHead = stuck.pop(0)
        curNode = newHead
        numNode = len(stuck)
        for i in range(numNode):
            curNode.next = stuck.pop(0)
            curNode = curNode.next
        curNode.next = None

        return newHead


# Sample Testcase:
#   Input:
#       1->2->3->4->5->NULL
#   Output:
#       5->4->3->2->1->NULL

# Performance Result:
#   Coding Time: 00:07:50
#   Time Complexity: O(n)
#   Space Complexity: O(n)
#   Runtime: 52 ms, faster than 9.29% of Python3
#       online submissions for Reverse Linked List.
#   Memory Usage: 15.1 MB, less than 31.82% of Python3
#       online submissions for Reverse Linked List.
